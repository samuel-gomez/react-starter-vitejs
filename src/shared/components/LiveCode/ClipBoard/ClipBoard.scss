@import '@axa-fr/react-toolkit-core/src/common/scss/core.scss';

$bgColorEditor: #1e1e1e;

.af-clipboard {
  position: absolute;
  top: 10px;
  right: 20px;
  display: flex;
  flex-direction: column;
  text-align: center;
  z-index: 2;
  .glyphicon {
    color: $color-white;
    font-size: 1rem;
    cursor: pointer;
    transition: text-shadow ease-in 0.25s;
  }
  &:hover {
    .glyphicon {
      color: $color-flamingo;
    }
  }
  &__message {
    color: $bgColorEditor;
    transition: color linear 0.5s;
    position: absolute;
    bottom: -21px;
    right: -16px;
    width: 57px;
  }
  &--clicked &__message {
    color: $color-white;
  }
}
