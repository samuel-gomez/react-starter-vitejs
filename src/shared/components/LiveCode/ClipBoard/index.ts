export { default, useClipBoard } from './ClipBoard';
export type { TReturnUseClipBoard, TClipBoard } from './ClipBoard';
