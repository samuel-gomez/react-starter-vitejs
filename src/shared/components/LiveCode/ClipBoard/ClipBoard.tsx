import { useCallback, useEffect, useRef, useState } from 'react';
import { Button } from '@axa-fr/react-toolkit-all';
import './ClipBoard.scss';

export const useClipBoard = ({ content }: { content: string }) => {
  const [clicked, setClicked] = useState(false);
  const timerRef = useRef<NodeJS.Timeout | null>(null);

  const onClick = useCallback(() => {
    setClicked(true);
    navigator.clipboard.writeText(content);
    timerRef.current = setTimeout(() => setClicked(false), 1000);
  }, [content]);
  useEffect(() => () => clearTimeout(timerRef.current as NodeJS.Timeout), []);

  return { onClick, clicked };
};

export type TReturnUseClipBoard = ReturnType<typeof useClipBoard>;

export type TClipBoard = { content: string; useClipBoardFn?: typeof useClipBoard };

const ClipBoard = ({ content, useClipBoardFn = useClipBoard }: TClipBoard) => {
  const { onClick, clicked } = useClipBoardFn({ content });
  const classClicked = clicked ? ' af-clipboard--clicked' : '';

  return (
    <Button aria-label="Copy to clipboard" className={`af-btn--circle af-clipboard${classClicked}`} onClick={onClick}>
      <i className="glyphicon glyphicon-copy" />
      <span className="af-clipboard__message">Copied !</span>
    </Button>
  );
};

export default ClipBoard;
