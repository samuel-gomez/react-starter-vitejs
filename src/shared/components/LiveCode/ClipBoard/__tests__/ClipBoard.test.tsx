import { act, render, renderHook } from '@testing-library/react';
import ClipBoard, { useClipBoard } from '..';

describe('<ClipBoard/>', () => {
  it('Should render ClipBoard', () => {
    const { asFragment } = render(<ClipBoard content="<p>info</p>" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render ClipBoard with clicked classModifier', () => {
    const useClipBoardFn = vi.fn().mockReturnValue({ click: vi.fn(), clicked: true });
    const { asFragment } = render(<ClipBoard content="<p>info</p>" useClipBoardFn={useClipBoardFn} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('useClipBoard', () => {
  const content = 'content';
  const writeText = vi.fn();
  Object.assign(navigator, {
    clipboard: {
      writeText,
    },
  });

  it('Should call writeText method of navigator.clipboard when onClick method of useClipBoard have been called with content', () => {
    const { result } = renderHook(() => useClipBoard({ content }));
    vi.spyOn(navigator.clipboard, 'writeText');
    vi.useFakeTimers();
    act(() => result.current.onClick());
    expect(result.current.clicked).toEqual(true);
    expect(writeText).toHaveBeenCalledWith(content);
    act(() => {
      vi.advanceTimersByTime(1000);
    });
    expect(result.current.clicked).toEqual(false);
  });
});
