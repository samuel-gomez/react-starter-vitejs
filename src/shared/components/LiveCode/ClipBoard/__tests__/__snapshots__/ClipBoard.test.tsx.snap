// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ClipBoard/> > Should render ClipBoard 1`] = `
<DocumentFragment>
  <button
    aria-label="Copy to clipboard"
    class="af-btn--circle af-clipboard"
    type="button"
  >
    <i
      class="glyphicon glyphicon-copy"
    />
    <span
      class="af-clipboard__message"
    >
      Copied !
    </span>
  </button>
</DocumentFragment>
`;

exports[`<ClipBoard/> > Should render ClipBoard with clicked classModifier 1`] = `
<DocumentFragment>
  <button
    aria-label="Copy to clipboard"
    class="af-btn--circle af-clipboard af-clipboard--clicked"
    type="button"
  >
    <i
      class="glyphicon glyphicon-copy"
    />
    <span
      class="af-clipboard__message"
    >
      Copied !
    </span>
  </button>
</DocumentFragment>
`;
